contrib/internal/httptrace/httptrace.go

```diff
@@ -26,17 +26,44 @@ var (
 	cfg = newConfig()
 )
 
+type inferredSpanCreatedCtxKey struct{}
+
+type FinishSpanFunc = func(status int, errorFn func(int) bool, opts ...tracer.FinishOption)
+
 // StartRequestSpan starts an HTTP request span with the standard list of HTTP request span tags (http.method, http.url,
 // http.useragent). Any further span start option can be added with opts.
-func StartRequestSpan(r *http.Request, opts ...ddtrace.StartSpanOption) (tracer.Span, context.Context) {
+func StartRequestSpan(r *http.Request, opts ...ddtrace.StartSpanOption) (tracer.Span, context.Context, FinishSpanFunc) {
 	// Append our span options before the given ones so that the caller can "overwrite" them.
 	// TODO(): rework span start option handling (https://github.com/DataDog/dd-trace-go/issues/1352)
 
 	var ipTags map[string]string
 	if cfg.traceClientIP {
 		ipTags, _ = httpsec.ClientIPTags(r.Header, true, r.RemoteAddr)
 	}
+
 	nopts := make([]ddtrace.StartSpanOption, 0, len(opts)+1+len(ipTags))
+	spanParentCtx, spanParentErr := tracer.Extract(tracer.HTTPHeadersCarrier(r.Header))
+	spanLinksCtx, spanLinksOk := spanParentCtx.(ddtrace.SpanContextWithLinks)
+
+	var inferredProxySpan tracer.Span = nil
+
+	if cfg.inferredProxyServicesEnabled {
+		inferredStartSpanOpts := make([]ddtrace.StartSpanOption, 0, 1)
+
+		if spanLinksOk {
+			inferredStartSpanOpts = append(inferredStartSpanOpts, tracer.WithSpanLinks(spanLinksCtx.SpanLinks()))
+		}
+
+		if inferredProxySpan = tryCreateInferredProxySpan(r.Header, spanParentCtx, inferredStartSpanOpts...); inferredProxySpan != nil {
+			r.Header.Del(ProxyHeaderSystem)
+			r.Header.Del(ProxyHeaderStartTimeMs)
+			r.Header.Del(ProxyHeaderPath)
+			r.Header.Del(ProxyHeaderHttpMethod)
+			r.Header.Del(ProxyHeaderDomain)
+			r.Header.Del(ProxyHeaderStage)
```

Comment:
I don't think the tracer should be modifying the request and removing headers (e.g. some middleware that wants to inspect the raw request headers would be impacted from this).
I feel the previous alternative of setting a boolean in the request context is safer than doing this cc @darccio .

Reply:
Agreed. The tracer shouldn't modify anything except for adding info.

Reply:
Reverted